internal/pkg/manifest/pipeline.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package manifest
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"text/template"
+
+	"github.com/fatih/structs"
+	"gopkg.in/yaml.v3"
+
+	"github.com/aws/amazon-ecs-cli-v2/internal/pkg/archer"
+	"github.com/aws/amazon-ecs-cli-v2/templates"
+)
+
+// Provider represents the possible source provider
+// for this pipeline.
+type Provider int
+
+// this is definitely not thread-safe but we'll deal with it
+// when there's a need
+var providerSpecificProperties map[Provider]interface{} = make(map[Provider]interface{})
+
+const (
+	// Github is the provider that sources from a Github repository.
+	Github Provider = iota + 1
+)
+
+// GithubProperties contain information to configure a Github source provider.
+type GithubProperties struct {
+	// use tag from https://godoc.org/github.com/fatih/structs#example-Map--Tags
+	// to specify the name of the field in the output properties
+	Repository string `structs:"repository"`
+}
+
+func (p Provider) String() string {
+	names := [...]string{
+		"uninitialized",
+		"github",
+	}
+
+	if int(p) < 0 || int(p) >= len(names) {
+		return "unknown"
+	}
+
+	return names[p]
+}
+
+// ConfiguredWith allows configuring the provider with provider-specific
+// properties.
+func (p Provider) ConfiguredWith(newProps interface{}) error {
```

Comment:
dumb question: What do you think if we have explicit structs for each provider?  
`GitHubProvider`, `CodeCommitProvider` and they all implement a `Name()` and `Properties()` interface?

I think this way we can have the constructor for each provider to take in required params and we can get rid of this method.

Reply:
Sure. Sigh...Go's enum